ThreadAPI: returned resp_total for existing question and new thread

wajeeha-khalid opened the pull request:
MA-1359 Thread GET: returned resp_total for existing question and new thread

wajeeha-khalid commented:
@nasthagiri @BenjiLee

nasthagiri commented:
Are there any existing unit tests for this that can be updated?

nasthagiri commented:
Does this address https://openedx.atlassian.net/browse/MA-1475?

BenjiLee commented:
That would be a nice two birds with one stone if it addresses MA-1475 :) 

Same question about the unit tests.

wajeeha-khalid commented:
@nasthagiri @BenjiLee 
I have updated PR to address MA-1475 too. Corresponding changes are also updated for discussion API at existing PR for MA-1359

Travis CI test were run for this PR and they are not failing by which it seems there is no specific test to compare `resp_total` values for both Thread GET and POST.
I am looking into ruby code; lets see if I can understand/update tests in short time.

wajeeha-khalid commented:
updated tests

BenjiLee commented:
This seems okay. Only qualm is that we are adding more arguments to the to_hash function as we might do in https://github.com/edx/cs_comments_service/pull/145/files

jcdyer commented:
I don't love the fact that we're special casing `is_new` to return a response count when we don't ask for responses.  It feels unclean.  On the other hand, the alternative would be to specify `with_responses` when we create a new object, which would mean extra db queries that we already know will return nothing.  A third option would be to have the `api/commentables.rb` POST function add  `resp_total` to the hash after it gets it back from the presenter, but then the view code starts taking over the presentation code's job, which is also not great.  Honestly, I don't love any of these solutions.  Anyone have any better ideas?

I also don't love the opaque positional arguments on everything: `(nil, false, 0, nil, true, "Micronesia", 0, 0)`, but that seems like it would be pretty invasive to try to change that.  (From five minutes of googling, it looks like arguments have to be either positional or keyword, unlike python, where any most arguments can be used either way).  Probably not worth the effort, especially if the plan is to move this api to python.

wajeeha-khalid commented:
@jcdyer About the arguments; I agree; if we are planning a revamp then its better to avoid any major refactoring till then.

About `resp_total=0`; I would like to avoid extra DB calls to have a cleaner code; I find adding it explicitly to hash on return to view, a better option. `to_hash` does not seem to have any idea what type of thread (new/existing) is calling it. @BenjiLee any ideas?

BenjiLee commented:
I could be wrong, but couldn't we just return `resp_total=0` if `resp_total` isn't set assuming that all to_hash calls require `resp_total`?

jcdyer commented:
@BenjiLee Not all `to_hash()` calls should return a `resp_total`.  Sometimes you just want an object without its children, and then it doesn't make the query to see how many children it has.  Returning resp_total=0 would be a lie.

BenjiLee commented:
I guess I was wrong!

My take on this would be to have unit tests to test the cases we are interested in, as well as the db calls and then have the `is_new` call.

wajeeha-khalid commented:
@BenjiLee unit test are already updated to test `resp_total` value where required.

jcdyer commented:
@wajeeha-khalid I think I agree with you.  The presenter is already taking too much responsibility. `is_new` should only ever be used in one place, where a new thread is being created.  If you want to add the `resp_total` to the hash after it is returned by the presenter and take away the `is_new` argument, I'd be happy with that.

wajeeha-khalid commented:
@jcdyer @BenjiLee updated

jcdyer commented:
Looks clean.  Looks good.  :+1:

BenjiLee commented:
:+1: Looks much neater o.O

jcdyer commented:
Nice work, @wajeeha-khalid.  Merge it.
